src/lib/utils/defer.ts

```diff
@@ -1,35 +1,48 @@
 const nextMicroTask = Promise.resolve();
 
 type Callback = (...args: any[]) => void;
-type Cancellable = Callback & {
+type Defer = Callback & {
+  wait(): Promise<void>;
   cancel(): void;
 };
 
-const defer = (callback: Callback): Cancellable => {
+const defer = (callback: Callback): Defer => {
   let progress: Promise<void> | null = null;
   let cancelled = false;
 
-  const fn: Cancellable = (...args) => {
+  const fn: Defer = (...args) => {
     if (progress !== null) {
       return;
     }
 
     progress = nextMicroTask.then(() => {
+      progress = null;
+
       if (cancelled) {
-        progress = null;
         cancelled = false;
         return;
       }
 
       callback(...args);
-      progress = null;
     });
   };
 
+  fn.wait = () => {
+    if (progress === null) {
```

Comment:
Do we need this check in production code? If not, we can wrap it in a `__DEV__` condition.

Reply:
WDYM? If we don't perform the check the the function will throw an unexpected the error.

Reply:
From what I understand, this check is only performed for us to never call `defer` before `wait`, right? I guess my reasoning is that this shouldn't happen at runtime in the user bundle so it might not be worth including it in the end bundle.

Reply:
I wouldn't care that much, it will save ~90 bytes. I can remove it and let it throw an error but that could bite us later I think.